pkg/autohealing/cloudprovider/openstack/provider.go

```diff
@@ -288,15 +289,40 @@ func (provider CloudProvider) firstTimeRepair(n healthcheck.NodeInfo, serverID s
 			// Uncordon the node
 			if n.IsWorker {
 				nodeName := n.KubeNode.Name
-				newNode := n.KubeNode.DeepCopy()
-				newNode.Spec.Unschedulable = false
-				if _, err := provider.KubeClient.CoreV1().Nodes().Update(context.TODO(), newNode, metav1.UpdateOptions{}); err != nil {
-					log.Errorf("Failed to cordon node %s, error: %v", nodeName, err)
-				} else {
-					log.Infof("Node %s is cordoned", nodeName)
+				// timeout for wait.Poll
+				ctx := context.Background()
+				errServiceUp := wait.PollUntilContextTimeout(ctx, 3*time.Second, provider.Config.RebuildDelayAfterReboot, false,
+					func(ctx context.Context) (bool, error) {
+						rapairedNode, getErr := provider.KubeClient.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+						if getErr != nil {
+							log.Errorf("Failed to get node %s, error: %v in wait.PollUntilContextTimeout waiting for services to start", nodeName, getErr)
+							return false, getErr
+						}
+						if strings.Contains(rapairedNode.Status.String(), "Type:Ready,Status:True") {
+							retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+								// Retrieve the latest version of Node before attempting update
+								// RetryOnConflict uses exponential backoff to avoid exhausting the apiserver
+								newNode, err := provider.KubeClient.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+								if err != nil {
+									log.Errorf("Failed to get node %s, error: %v before update", nodeName, err)
+								}
+								newNode.Spec.Unschedulable = false
+								if _, updateErr := provider.KubeClient.CoreV1().Nodes().Update(context.TODO(), newNode, metav1.UpdateOptions{}); updateErr != nil {
+									log.Errorf("Failed to uncordon node %s, error: %v", nodeName, updateErr)
+									return updateErr
```

Comment:
If it's a retriable error I'd just log a warning instead of error.

Reply:
Changed